staging/src/k8s.io/apiserver/pkg/endpoints/filters/impersonation/mode.go

```diff
@@ -0,0 +1,628 @@
+		return true
+	}
+	var count int
+	for _, values := range extra {
+		count += len(values)
+		if count >= manyAuthorizationChecksInLoop {
+			return true
+		}
+	}
+	return false
+}
+
+func impersonationAttributes(requestor user.Info, gv schema.GroupVersion, verb, resource, name string) authorizer.AttributesRecord {
+	return authorizer.AttributesRecord{
+		User:            requestor,
+		Verb:            verb,
+		APIGroup:        gv.Group,
+		APIVersion:      gv.Version,
+		Resource:        resource,
+		Name:            name,
+		ResourceRequest: true,
+	}
+}
+
+// impersonateOnAttributes is a simple wrapper that updates the verb of the attributes to impersonate-on:<mode>:<verb>
+// This allows the expression of "a subject can perform this verb while using this impersonation mode"
+type impersonateOnAttributes struct {
+	mode string
+	authorizer.Attributes
+}
+
+func (i *impersonateOnAttributes) GetVerb() string {
+	return "impersonate-on:" + i.mode + ":" + i.Attributes.GetVerb()
+}
+
+func checkAuthorization(ctx context.Context, a authorizer.Authorizer, attributes authorizer.Attributes) error {
+	authorized, reason, err := a.Authorize(ctx, attributes)
+
+	// an authorizer like RBAC could encounter evaluation errors and still allow the request, so authorizer decision is checked before error here.
+	if authorized == authorizer.DecisionAllow {
+		return nil
+	}
+
+	// if the authorizer gave us a forbidden error, do not wrap it again
+	if errors.IsForbidden(err) {
+		return err
+	}
+
+	msg := reason
+	switch {
+	case err != nil && len(reason) > 0:
+		msg = fmt.Sprintf("%v: %s", err, reason)
+	case err != nil:
+		msg = err.Error()
+	}
+
+	return responsewriters.ForbiddenStatusError(attributes, msg)
+}
+
+func ensureGroup(u *user.DefaultInfo, group string) {
+	if slices.Contains(u.Groups, group) {
+		return
+	}
+
+	// do not mutate a slice that we did not create
+	groups := make([]string, 0, len(u.Groups)+1)
+	groups = append(groups, u.Groups...)
+	groups = append(groups, group)
+	u.Groups = groups
+}
+
+func isServiceAccountUsername(username string) (namespace, name string, ok bool) {
+	namespace, name, err := serviceaccount.SplitUsername(username)
+	return namespace, name, err == nil
+}
+
+// matches the real ValidateNodeName from k8s.io/kubernetes/pkg/apis/core/validation
+// which we are not allowed to import here
+var validateNodeName = validation.NameIsDNSSubdomain
+
+func isNodeUsername(username string) (string, bool) {
+	const nodeUsernamePrefix = "system:node:"
+	if !strings.HasPrefix(username, nodeUsernamePrefix) {
+		return "", false
+	}
+	name := strings.TrimPrefix(username, nodeUsernamePrefix)
+	if len(validateNodeName(name, false)) != 0 {
+		return "", false
+	}
+	return name, true
+}
+
+func requesterAssociatedWithNode(requestor user.Info, username string) bool {
+	nodeName, ok := isNodeUsername(username)
+	if !ok {
+		return false
+	}
+	if _, _, ok := isServiceAccountUsername(requestor.GetName()); !ok {
+		return false
+	}
```

Comment:
do we care if the requester is a serviceaccount, or just that the node-name shows up in the extra info?